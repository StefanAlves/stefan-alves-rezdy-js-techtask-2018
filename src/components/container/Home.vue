<!-- src/components/HelloDecorator.vue -->
<!-- This is an alternative way to define the Hello component using decorators -->
<template>
  <div>
    <button class="btn" @click="displayRecipes">What's for Lunch</button>
    <div v-if="showRecipes" :recipes="getRecipes">
      <recipes :recipes="getRecipes" />
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";
import { mapGetters } from 'vuex'
import Recipes from '../presentational/Recipes.vue';

@Component({
  components: { Recipes },
  computed: mapGetters([
    'getRecipes'
  ])
})
export default class Home extends Vue {
  @Prop() recipes!: string;
  // local props
  showRecipes = false;
  /**
   * show recipes
   */
  displayRecipes() {
    this.showRecipes = true;
  }
}
</script>

<style>
  .btn{
    background-color: #4CAF50; /* Green */
    border: none;
    color: white;
    padding: 15px 32px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;
    width: 100%;
  }
</style>