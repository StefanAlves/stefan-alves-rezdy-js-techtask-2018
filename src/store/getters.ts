import { State, Ingredient, Recipe } from '../types';

/**
 * Filter recipes based on used by date
 * @param recipes: Array of Recipe 
 * @param ingredients: Map of Ingredients
 * @param date - Timestamp 
 * @returns Array recipes
 */
export const filterRecipes = (recipes: Recipe[], ingredients: { [key: string]: Ingredient}, beforeDate: number) => {
  return recipes.filter((recipe: Recipe) => {
    const hasAnyInvalid = recipe.ingredients.filter((ingredientTitle: string) => {
      const ingredient: Ingredient = ingredients[ingredientTitle];
      if(ingredient){
        const usedBy = Date.parse(ingredient['use-by']);
        return beforeDate > usedBy;
      }
      return false;
    });
    return hasAnyInvalid.length === 0; 
  })
}

/**
 * Sort recipes based on used best before
 * @param recipes: Array of Recipe 
 * @param ingredients: Map of Ingredients
 * @param date - Timestamp
 * @returns Array recipes 
 */
export const sortRecipes = (recipes: Recipe[], ingredients: { [key: string]: Ingredient}, beforeDate: number) => {
  const sortedRecipes: Recipe[] = [];
  /**
   * Keep previous order so let's reverse the foreach 
   */
  for (let index = recipes.length-1; index >= 0; index--) {
    const recipe = recipes[index];
    const hasAnyBestBefore = recipe.ingredients.filter((ingredientTitle: string) => {
      const ingredient: Ingredient = ingredients[ingredientTitle];
      if(ingredient){
        const usedBy = Date.parse(ingredient['best-before']);
        return beforeDate > usedBy;
      }
      return false;
    });
    if(hasAnyBestBefore.length > 0){
      sortedRecipes.push(recipe);
    }else{
      sortedRecipes.unshift(recipe);
    }
  }
  return sortedRecipes;
}

export default {
  /**
   * get app recipes
   * @returns Array recipes
   */
  getRecipes: (state: State ) => {
    const { recipes, ingredients } = state;
    const today = Date.now(); 
    const filteredRecipes = filterRecipes(recipes, ingredients, today);
    const sorteRecipes = sortRecipes(filteredRecipes, ingredients, today);
    return sorteRecipes;
  },
}