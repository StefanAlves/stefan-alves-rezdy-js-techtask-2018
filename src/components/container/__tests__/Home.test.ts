import { mount, createLocalVue, config } from '@vue/test-utils';
import * as Vuex from 'vuex';
import getters from '../../../store/getters';
import Home from '../Home.vue';

const localVue = createLocalVue();
localVue.use(Vuex);
config.logModifiedComponents = false;

describe('Home', () => {

  let store: any;
  
  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        recipes : [
          {
            title: 'Fry-up',
            ingredients: [
                'Bacon',
                'Eggs'
            ]
          },
          {
            title: 'Salad',
            ingredients: [
              'Lettuce',
              'Tomato'
            ]
          }
        ],
        ingredients: {
          Bacon: {
            title: 'Bacon',
            'best-before': '2018-11-25',
            'use-by': '2018-11-30'
          },
          Eggs: {
            title: 'Eggs',
            'best-before': '2018-11-20',
            'use-by': '2018-11-25'
          },
          Lettuce: {
            title: 'Lettuce',
            'best-before': '2018-11-20',
            'use-by': '2018-11-25'
          },
          Tomato: {
            title: 'Tomato',
            'best-before': '2018-11-20',
            'use-by': '2018-11-25'
          }
        }
      },
      getters
    });
  })

  it('should render', () => {

  // localVue.use(Vuex)
    const wrapper = mount(Home, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
    const button = wrapper.find('button')
    button.trigger('click');
    expect(wrapper.element).toMatchSnapshot();
    // console.log(wrapper.html());
    // const recipeStubs = wrapper.findAll('recipe-stub');
    // expect(recipeStubs.length).toEqual(2)
  });
});