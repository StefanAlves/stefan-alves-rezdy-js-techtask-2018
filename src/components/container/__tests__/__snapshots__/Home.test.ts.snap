// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render 1`] = `
<div>
  <button
    class="btn"
  >
    What's for Lunch
  </button>
   
  <!---->
</div>
`;

exports[`Home should render 2`] = `
<div>
  <button
    class="btn"
  >
    What's for Lunch
  </button>
   
  <div
    recipes="[object Object],[object Object]"
  >
    <div>
      <div
        class="recipe"
      >
        <div
          class="title"
        >
          Fry-up
        </div>
         
        <ul
          class="ingredients"
        >
          <li
            class="ingredient"
          >
            
      Bacon
    
          </li>
          <li
            class="ingredient"
          >
            
      Eggs
    
          </li>
        </ul>
      </div>
      <div
        class="recipe"
      >
        <div
          class="title"
        >
          Salad
        </div>
         
        <ul
          class="ingredients"
        >
          <li
            class="ingredient"
          >
            
      Lettuce
    
          </li>
          <li
            class="ingredient"
          >
            
      Tomato
    
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
